akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala

```diff
@@ -49,8 +50,16 @@ abstract class ActorRefResolver extends Extension {
 
   private val untypedSystem = system.toUntyped.asInstanceOf[ExtendedActorSystem]
 
-  override def toSerializationFormat[T](ref: ActorRef[T]): String =
-    ref.path.toSerializationFormatWithAddress(untypedSystem.provider.getDefaultAddress)
+  override def toSerializationFormat[T](ref: ActorRef[T]): String = {
+    val originalSystem: ExtendedActorSystem = ref.toUntyped match {
+      case a: ActorRefWithCell => a.underlying.system.asInstanceOf[ExtendedActorSystem]
+      case _                   => null
+    }
+    originalSystem match {
+      case null     => ref.path.toSerializationFormatWithAddress(untypedSystem.provider.getDefaultAddress)
+      case original => ref.path.toSerializationFormatWithAddress(original.provider.getDefaultAddress)
+    }
```

Comment:
Hmm, so we have different behavior based on whether we can find the system based on the `ActorRef` or not... if people start to depend on such logic I fear it'd be hard to refactor later.

Perhaps we should just more strongly protect against sharing `ActorRef` instances between systems, and throw an error when the system doesn't match?

Reply:
Yep I think that is a better idea :+1:

Reply:
> protect against sharing `ActorRef` instances between systems

WDYM? Isn't it a feature that you can use ActorRefs between local ActorSystems?

Reply:
> Hmm, so we have different behavior based on whether we can find the system based on the `ActorRef` or not... if people start to depend on such logic I fear it'd be hard to refactor later.

I guess the question is what the `case _ => null` comprises exactly. If it's only remote actor refs then it shouldn't matter, if there's anything else in there, we should probably try harder to get the correct system.

Reply:
> Isn't it a feature that you can use ActorRefs between local ActorSystems?

It's wrong to pass a local ActorRef by reference from one system to another system. Sometimes it just works anyway but sometimes not. Such refs should always be retrieved in the same way as if the other ActorSystem was remote, i.e. they will be RemoteActorRef.

I think `MinimalActorRef` doesn't have a cell so it falls into the `null` case.

Comparing to `Serialization.serializedActorPath` that one is also looking at the ThreadLocal currentTransportInformation, but on the other hand it doesn't have a `system` like this extension has. I think it's less magic to use the `system` of the extension as the fallback.

Since we can't (without changes that are hard to motivate) support all types of refs (MinimalActorRef) I'm also leaning towards that we should instead disallow refs belonging to the wrong system (via the cell access).